### ISSUE TITLE
ReferenceError: head is not defined

### ISSUE THREAD

**Issue by cressie176** [peggyjs/peggy#275]
Hello & thank you for your work on peggy and for picking up where pegjs left off. 

I'm one of the maintainers of [amqplib](https://github.com/amqp-node/amqplib) and investigating https://github.com/amqp-node/amqplib/issues/453. amqplib uses pegjs via [bitsyntax](https://github.com/squaremo/bitsyntax-js) to generate a parser from the grammar specified [here](https://github.com/squaremo/bitsyntax-js/blob/master/lib/grammar.pegjs).

bitsyntax's last working parser was generated using pegjs@0.7.0 and upgrading to pegjs@0.1.0, peggy@^1.0.0 or peggy@latest results in a broken parser.

```
/Users/steve/Development/squaremo/bitsyntax-js/lib/parser.js:256
  var peg$f17 = function() { return head + tail.join(''); };
                             ^

ReferenceError: head is not defined
```

I'm completely new to pegjs/peggy and was hoping it's an easy fix. Can you give me any pointers please?

Thanks

**Comment by hildjj** [comment]
I grabbed the grammar, ran it through peggy's CLI with the default options, then called `parse('"foo", "bar"')`, and got `[ { string: 'foo' }, { string: 'bar' } ]`.  I suspect the issue may be [here](https://github.com/squaremo/bitsyntax-js/blob/16622ff739d947eed0d2fb97cf22eeacf83450c0/Makefile#L9).  Try making that like be:

```makefile
	 $(PEGJS) $(GRAMMAR) -o $@
```
 
(add `-o` to direct the output to a file).  My suspicion is that you're getting an error instead of generating the new implementation.  Assuming it's not that, can you tell us how you're calling peggy, and an example input that's failing for you?

I agree that the line you've highlighted looks wrong.  Here's what I get:

```js
var peg$f0 = function(head, tail) { tail.unshift(head); return tail; };
```

**Comment by cressie176** [comment]
Hi @hildjj,

Thanks for the quick response. I also noticed that the Makefile is out of date. I've generated the parser by hand as follows

```
rm lib/parser.js 
./node_modules/.bin/peggy -o lib/parser.js lib/grammar.pegjs
```

**Comment by hildjj** [comment]
Example input, please?  And can you check that you're using Peggy 2.0.1, which was just released yesterday?

**Comment by hildjj** [comment]
OK, it has nothing to do with the build process.  Remove the parens on line 43:

```peg.js
identifier
    = head:[_a-zA-Z] tail:[_a-zA-Z0-9]* { return head + tail.join(''); }
```

which ends up being `"a" + ["b", "c"]` which is a roundabout way of getting the string that matched.  You could just use:

```peg.js
identifier
    = [_a-zA-Z] [_a-zA-Z0-9]* { return text(); }
```

Some old version of peg.js must have changed how parens are handled.  In today's world, they hide labels from the containing rule.  Once I got that working, there are a bunch of other test errors, but that's a separate issue.

**Comment by cressie176** [comment]
Thanks @hildjj. I'll close this, and try to work out why the tests are failing.

**Comment by hildjj** [comment]
Let us know if you'd like more help with the grammar.  Happy to pitch in as needed.

**Comment by Mingun** [comment]
Actually, your original grammar works because of [a bug](https://github.com/pegjs/pegjs/blob/master/docs/changelogs/v0.10.md), [fixed](0c39f1cf86d73c85a1f832e6cfddf4176f9e58bc) in pegjs@0.10.0:
> Fixed label scoping so that labels in expressions like `(a:"a")` or `(a:"a" b:"b" c:"c")` aren’t visible from the outside.

Also, you could simplify your grammar much more using [a new `@` operator](https://peggyjs.org/documentation.html#grammar-syntax-and-semantics-parsing-expression-types), for example, here:
```pegjs
segmentTail
    = ws ',' ws @segment
```
and using a `$` operator instead of `return text()`:
```pegjs
identifier
    = $([_a-zA-Z] [_a-zA-Z0-9]*)
```
